packages/superset-ui-chart/src/models/ChartPlugin.ts

```diff
@@ -0,0 +1,79 @@
+import { isRequired, Plugin } from '@superset-ui/core';
+import { ChartMetadata } from './ChartMetadata';
+import { ChartProps } from './ChartProps';
+import { FormData } from '../query/FormData';
+import { QueryContext } from '../query/buildQueryContext';
+import getChartMetadataRegistry from '../registries/ChartMetadataRegistrySingleton';
+import getChartBuildQueryRegistry from '../registries/ChartBuildQueryRegistrySingleton';
+import getChartComponentRegistry from '../registries/ChartComponentRegistrySingleton';
+import getChartTransformPropsRegistry from '../registries/ChartTransformPropsRegistrySingleton';
+
+const IDENTITY = (x: any) => x;
+
+type PromiseOrValue<T> = Promise<T> | T;
+type PromiseOrValueLoader<T> = () => PromiseOrValue<T>;
+
+export type BuildQueryFunction = (formData: FormData) => QueryContext;
+
+export type TransformPropsFunction = (
+  chartProps: ChartProps,
+) => {
+  [key: string]: any;
+};
+
+export interface ChartPluginConfig {
+  metadata: ChartMetadata;
+  // use buildQuery for immediate value
+  buildQuery?: BuildQueryFunction;
+  // use loadBuildQuery for dynamic import (lazy-loading)
+  loadBuildQuery?: PromiseOrValueLoader<BuildQueryFunction>;
+  // use transformProps for immediate value
+  transformProps?: TransformPropsFunction;
+  // use loadTransformProps for dynamic import (lazy-loading)
+  loadTransformProps?: PromiseOrValueLoader<TransformPropsFunction>;
+  // use Chart for immediate value
+  Chart?: Function;
+  // use loadChart for dynamic import (lazy-loading)
+  loadChart?: PromiseOrValueLoader<Function>;
+}
+
+export class ChartPlugin extends Plugin {
+  metadata: ChartMetadata;
+  loadBuildQuery: PromiseOrValueLoader<BuildQueryFunction> | null;
+  loadTransformProps: PromiseOrValueLoader<TransformPropsFunction>;
+  loadChart: PromiseOrValueLoader<Function>;
+
+  constructor(config: ChartPluginConfig) {
+    super();
+    const {
+      metadata,
+      buildQuery,
+      loadBuildQuery,
+      transformProps = IDENTITY,
+      loadTransformProps,
+      Chart,
+      loadChart,
+    } = config;
+    this.metadata = metadata;
+    this.loadBuildQuery = loadBuildQuery || (buildQuery ? () => buildQuery : null);
```

Comment:
not sure if we should be consistent with `null`s vs `undefined`. previously I was a fan of `null`s but our TS style guide says

> Use undefined in general
>   *Why?* Compatible with optionals (the ? operator). Triggers default argument initializers. 
> Use null where its a part of the API or conventional
>  e.g., optional wouldn't be appropriate if you know a server will respond with a string or null / won't be omitted

Reply:
Good point. @conglei added this part but I think either `undefined` or `null` would work for him. Will change to `undefined`